### PR TITLE
use bake method for NavigationMeshGenerator in Runtime

### PR DESCRIPTION
**Pull request by thoced**
Use bake with NavigationMeshGenerator in the Runtime in Godot 3.x (Windows, Android)

Hello

The use of the Bake in **NavigationMeshGenerator** method did not work in runtime (because of EditorProgress which wants to display a message in runtime) but only in the editor (hint_editor). This fix allows you to generate navigationmesh directly in the runtime. This is very useful, in particular, **when creating procedural or random level.**

Now, **NavigationMeshGenerator** can be used in the editor and in the Runtime for Plateform **Windows** and **Android**
(with and without flag tool=no or tool=yes)

Thanks

### CODE REVIEW

File: modules/recast/navigation_mesh_generator.h [hunk 1]
```diff
@@ -50,9 +52,19 @@ class EditorNavigationMeshGenerator : public Object {
...
 
 	static void _convert_detail_mesh_to_native_navigation_mesh(const rcPolyMeshDetail *p_detail_mesh, Ref<NavigationMesh> p_nav_mesh);
-	static void _build_recast_navigation_mesh(Ref<NavigationMesh> p_nav_mesh, EditorProgress *ep,
+
+#ifdef TOOLS_ENABLED
+	static void _build_recast_navigation_mesh(Ref<NavigationMesh> p_nav_mesh,
+			rcHeightfield *hf, rcCompactHeightfield *chf, rcContourSet *cset, rcPolyMesh *poly_mesh,
+			rcPolyMeshDetail *detail_mesh, Vector<float> &vertices, Vector<int> &indices , EditorProgress *ep);
+#endif
+#ifndef TOOLS_ENABLED
```
Comment (LoipesMas) [inline modules/recast/navigation_mesh_generator.h:61]: Shouldn't this be just `#else` instead?

File: modules/recast/register_types.cpp [hunk 2]
```diff
@@ -30,32 +30,37 @@
 
 #include "register_types.h"
 
+#ifdef TOOLS_ENABLED
 #include "navigation_mesh_editor_plugin.h"
+#endif
+
+#ifndef TOOLS_ENABLED
```
Comment (LoipesMas) [inline modules/recast/register_types.cpp:37]: Same here

File: modules/recast/navigation_mesh_generator.h (hunk 1 above)
Comment (thoced) [reply modules/recast/navigation_mesh_generator.h]: Hello,
#else added
Thanks

**Comment by thoced** [comment]
Hello, can you tell me what action I still need to take for this code to be verified ?

**Comment by lexferreira89** [comment]
It would be great to have this feature in Godot 3.x. Without the ability to bake navigation meshes during runtime, it's impossible to create certain kinds of games with Godot (RTS, City Builders, Simulation, etc). While we wait for Godot 4 and its new navigation system, allowing the bake function to be used during runtime would be a game changer. So could someone please take a look at this?

File: modules/recast/navigation_mesh_generator.cpp [hunk 3]
```diff
@@ -294,11 +296,22 @@ void EditorNavigationMeshGenerator::_convert_detail_mesh_to_native_navigation_me
...
+#else
+void EditorNavigationMeshGenerator::_build_recast_navigation_mesh(Ref<NavigationMesh> p_nav_mesh,
+			rcHeightfield * hf, rcCompactHeightfield * chf, rcContourSet * cset, rcPolyMesh * poly_mesh, rcPolyMeshDetail * detail_mesh,
+			Vector<float> & vertices, Vector<int> & indices) {
+#endif
+
+rcContext ctx;
+
+#ifdef TOOLS_ENABLED
+	if (ep)
```
Comment (Calinou) [inline modules/recast/navigation_mesh_generator.cpp:312]: To follow the code style guidelines, please replace all instances of `if(ep)` with:

```
if (ep) {
```

Comment (Calinou) [inline modules/recast/navigation_mesh_generator.cpp:312]: To follow the code style guidelines, please replace all instances of `if(ep)` with:

```cpp
if (ep) {
```

Braces should now always be used for single-line `if` conditions.

**Review by Calinou — CHANGES_REQUESTED** [review]

**Comment by Calinou** [comment]
While this change is useful, remember that it will increase the export templates' binary size due to the Recast module now being compiled in export templates. Binary sizes are especially important on mobile/web platforms, so users will have to compile custom export templates if they wish to keep the reduced binary sizes.

**Comment by thoced** [comment]
Hello, Braces added in the code

**Comment by thoced** [comment]
Hello, the test has not passed yet, could someone help me figure out what is missing in the code to pass the static checks?
 thank you

**Comment by Calinou** [comment]
Copying from the CI log:

```diff
diff --git a/modules/recast/navigation_mesh_generator.cpp b/modules/recast/navigation_mesh_generator.cpp
index 10c9e2c..2d32ad1 100644
--- a/modules/recast/navigation_mesh_generator.cpp
+++ b/modules/recast/navigation_mesh_generator.cpp
@@ -297,7 +297,7 @@ void EditorNavigationMeshGenerator::_convert_detail_mesh_to_native_navigation_me
 }
 
 #ifdef TOOLS_ENABLED
-void EditorNavigationMeshGenerator::_build_recast_navigation_mesh(Ref<NavigationMesh> p_nav_mesh, 
+void EditorNavigationMeshGenerator::_build_recast_navigation_mesh(Ref<NavigationMesh> p_nav_mesh,
 		rcHeightfield *hf, rcCompactHeightfield *chf, rcContourSet *cset, rcPolyMesh *poly_mesh, rcPolyMeshDetail *detail_mesh,
 		Vector<float> &vertices, Vector<int> &indices , EditorProgress *ep = nullptr) {
 #else
@@ -462,7 +462,7 @@ void EditorNavigationMeshGenerator::_build_recast_navigation_mesh(Ref<Navigation
 	if (ep) {
 		ep->step(TTR("Converting to native navigation mesh..."), 10);
 	}
-#endif 
+#endif
 
 	_convert_detail_mesh_to_native_navigation_mesh(detail_mesh, p_nav_mesh);
 
@@ -492,11 +492,11 @@ void EditorNavigationMeshGenerator::bake(Ref<NavigationMesh> p_nav_mesh, Node *p
 		ep = memnew(EditorProgress("bake", TTR("Navigation Mesh Generator Setup:"), 11));
 		ep->step(TTR("Parsing Geometry..."), 0);
 	}
-#endif 
+#endif
 
 
 	ERR_FAIL_COND(!p_nav_mesh.is_valid());
-		
+
 	Vector<float> vertices;
 	Vector<int> indices;
 
diff --git a/modules/recast/navigation_mesh_generator.h b/modules/recast/navigation_mesh_generator.h
index afa919c..c3b9162 100644
--- a/modules/recast/navigation_mesh_generator.h
+++ b/modules/recast/navigation_mesh_generator.h
@@ -61,7 +61,7 @@ protected:
 	static void _build_recast_navigation_mesh(Ref<NavigationMesh> p_nav_mesh,
 			rcHeightfield *hf, rcCompactHeightfield *chf, rcContourSet *cset, rcPolyMesh *poly_mesh,
 			rcPolyMeshDetail *detail_mesh, Vector<float> &vertices, Vector<int> &indices);
-#endif 
+#endif
 
 public:
 	static EditorNavigationMeshGenerator *get_singleton();

*** Aborting, please fix your commit(s) with 'git commit --amend' or 'git rebase -i <hash>'
```

You need to remove trailing whitespace at the end of some lines. I recommend configuring your editor to do so automatically.

**Comment by lexferreira89** [comment]
Hi! I just wanted to thank thoced for all his work in this PR. Do you guys think it's still possible for this feature to be included in the next release? Sorry if I'm being annoying!

**Comment by thoced** [comment]
Hello everyone
What is left for me to do now?
Thank you

**Comment by LoipesMas** [comment]
As Calinou mentioned, all commits must be squashed to a single one before the PR can be merged. 
[Here are the instructions on how to do it. ](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase)

**Comment by akien-mga** [comment]
> "Cannot rebase: You have unstaged changes.
Please commit or stash them. "

This means that you have local changes to files which have not been committed. You either need to commit them, or discard them, before you can do operations on the Git history. You can also "stash them" with `git stash`, which puts them away, and you can recall them with `git stash pop` later on.

You can see what changes you have pending with `git status` and `git diff`. See https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#making-changes for some Git tips.

**Comment by Calinou** [comment]
Superseded by https://github.com/godotengine/godot/pull/54634.

**Pull request closed without merging**